src/chain/chain-generic-numerator.h

```diff
@@ -81,90 +100,94 @@ namespace chain {
-  // setting it small is that we have to invoke an AddMat kernel more times.
-  enum { kMaxDerivTimeSteps = 8 };
+  // For the remapped FSTs, copy the appropriate activations to CPU memory.
+  // For explanation of what remapped FST is, see the large comment in the
+  // beginning of the file
+  void CopySpecificPdfsIndirect(int sequence_id,
+                             const CuMatrixBase<BaseFloat> &nnet_output,
+                             const std::vector<MatrixIndexT> &indices,
+                             Matrix<BaseFloat> *output);
+
+  // For the remapped FSTs, copy the computed values back to gpu,
+  // expand to the original shape and add to the output matrix.
+  // For explanation of what remapped FST is, see the large comment in the
+  // beginning of the file.
+  void AddSpecificPdfsIndirect(int sequence_id,
+                             const Matrix<BaseFloat> &logprobs,
+                             const std::vector<MatrixIndexT> &indices,
+                             CuMatrixBase<BaseFloat> *output);
 
   // sets up the alpha for frame t = 0.
-  void AlphaFirstFrame();
-
-  // the alpha computation for some 0 < t <= num_time_steps_.
-  void AlphaGeneralFrame(int32 t);
-
-  BaseFloat ComputeTotLogLike();
-
-  // sets up the beta for frame t = num_time_steps_.
-  void BetaLastFrame();
-
-  // the beta computation for 0 <= beta < num_time_steps_.
-  void BetaGeneralFrame(int32 t);
+  void AlphaFirstFrame(int seq, Matrix<BaseFloat> *alpha);
+
+  // the alpha computation for 0 < t <= supervision_.frames_per_sequence
+  // for some 0 <= seq < supervision_.num_sequences.
+  BaseFloat AlphaGeneralFrame(int seq,
+                              const Matrix<BaseFloat> &probs,
+                              Matrix<BaseFloat> *alpha);
+
+  // the beta computation for 0 <= t < supervision_.frames_per_sequence
+  // for some 0 <= seq < supervision_.num_sequences.
+  void BetaGeneralFrame(int32 seq,
+                        const Matrix<BaseFloat> &probs,
+                        const Matrix<BaseFloat> &alpha,
+                        Matrix<BaseFloat> *beta,
+                        Matrix<BaseFloat> *derivs);
+
+  // the beta computation for t = supervision_.frames_per_sequence
+  void BetaLastFrame(int seq,
+                     const Matrix<BaseFloat> &probs,
+                     const Matrix<BaseFloat> &alpha,
+                     Matrix<BaseFloat> *beta,
+                     Matrix<BaseFloat> *derivs);
+
+  // returns total prob for the given matrix alpha (assumes the alpha
+  // matrix was computed using AlphaFirstFrame() and AlphaGeneralFrame()
+  // (it's exactly like 'tot_probe_' in DenominatorComputation)
+  BaseFloat GetTotalProb(const Matrix<BaseFloat> &alpha);
 
   // some checking that we can do if debug mode is activated, or on frame zero.
   // Sets ok_ to false if a bad problem is detected.
-  void BetaGeneralFrameDebug(int32 t);
+  bool BetaGeneralFrameDebug(int32 seq,
```

Comment:
This can be renamed to something like CheckValues(); and make it const if it doesn't change any class members.